kin-sdk-core/src/androidTest/java/kin/sdk/core/RequestTest.java

```diff
@@ -0,0 +1,196 @@
+            }
+
+            @Override
+            public void onError(Exception e) {
+                if (onErrorCallback != null) {
+                    onErrorCallback.accept(e);
+                }
+                latch.countDown();
+            }
+        });
+        assertTrue(latch.await(TIMEOUT_DURATION_MILLIS, TimeUnit.MILLISECONDS));
+    }
+
+    @Test
+    public void run_verifyExpectedResult() throws InterruptedException {
+        String expectedResult = "ExpectedResult";
+        runRequest(
+            () -> expectedResult,
+            result -> assertEquals(expectedResult, result),
+            null
+        );
+    }
+
+    @Test
+    public void run_verifyCorrectThreads() throws InterruptedException {
+        AtomicLongArray threadIds = new AtomicLongArray(2);
+        runRequest(() -> {
+                threadIds.set(0, Thread.currentThread().getId());
+                return null;
+            },
+            result -> threadIds.set(1, Thread.currentThread().getId()),
+            null
+        );
+
+        long mainThreadId = Looper.getMainLooper().getThread().getId();
+        assertNotEquals(mainThreadId, threadIds.get(0));
+        assertEquals(mainThreadId, threadIds.get(1));
+    }
+
+    @Test
+    public void run_verifyExceptionPropagation() throws InterruptedException {
+        Exception expectedException = new Exception("some exception");
+        runRequest(() -> {
+                throw expectedException;
+            },
+            null
+            , e -> assertEquals(expectedException, e));
+    }
+
+    @Test
+    public void run_cancelBeforeRun() throws InterruptedException {
+        CountDownLatch latch = new CountDownLatch(1);
+        AtomicInteger integer = new AtomicInteger(0);
+        Request<Object> dummyRequest = new Request<>(() -> {
+            integer.incrementAndGet();
+            return new Object();
+        });
+        dummyRequest.run(new ResultCallback<Object>() {
+            @Override
+            public void onResult(Object result) {
+                integer.incrementAndGet();
+            }
+
+            @Override
+            public void onError(Exception e) {
+                integer.incrementAndGet();
+            }
+        });
+        dummyRequest.cancel();
+        latch.await(TIMEOUT_DURATION_MILLIS, TimeUnit.MILLISECONDS);
+        assertEquals(0, integer.get());
+    }
+
+    @Test
+    public void run_cancelAfterRun() throws InterruptedException {
+        CountDownLatch latch = new CountDownLatch(1);
+        AtomicBoolean threadInterrupted = new AtomicBoolean(false);
+        Request<Object> dummyRequest = new Request<>(() -> {
+            try {
+                Thread.sleep(TASK_DURATION_MILLIS);
+            } catch (InterruptedException ie) {
+                threadInterrupted.set(true);
+                latch.countDown();
+            }
+            return new Object();
+        });
+        dummyRequest.run(new ResultCallback<Object>() {
```

Comment:
we should not check whether onResult/onError have been called?

Reply:
indeed, should check it hasn't been called, test enhanced